src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala

```diff
@@ -495,24 +498,41 @@ class GroupReadsByUmi
     val out = SamWriter(output, outHeader)
 
     val iterator = Bams.templateIterator(sorter.iterator, out.header, Bams.MaxInMemory, Io.tmpDir)
-    val tagFamilySizeCounter = new NumericCounter[Int]()
-
-    while (iterator.hasNext) {
-      // Take the next set of templates by position and assign UMIs
-      val templates = takeNextGroup(iterator)
-      assignUmiGroups(templates)
-
-      // Then output the records in the right order (assigned tag, read name, r1, r2)
-      val templatesByMi = templates.groupBy { t => t.r1.get[String](this.assignTag) }
 
-      templatesByMi.keys.toSeq.sortBy(id => (id.length, id)).foreach(tag => {
-        templatesByMi(tag).sortBy(t => t.name).flatMap(t => t.primaryReads).foreach(rec => {
-          out += rec
-        })
-      })
+    val tagFamilySizeCounters = new IterableThreadLocal(() => new NumericCounter[Int]())
+    val pool                  = new ForkJoinPool(threads, ForkJoinPool.defaultForkJoinWorkerThreadFactory, null, true)
+
+    Iterator.continually(if (iterator.hasNext) takeNextGroup(iterator) else Seq.empty)
+      .takeWhile(_.nonEmpty)
+      .grouped(100000)
+      .foreach { group =>
+        group
+          .parWith(pool)
+          .map { templates: Seq[Template] =>
+            // Take the next set of templates by position and assign UMIs
+            assignUmiGroups(templates)
+
+            // Then output the records in the right order (assigned tag, read name, r1, r2)
+            val templatesByMi = templates.groupBy { t => t.r1.get[String](this.assignTag) }
+
+            // Count up the family sizes
+            val tagFamilySizeCounter = tagFamilySizeCounters.get()
+            templatesByMi.values.foreach(ps => tagFamilySizeCounter.count(ps.size))
+
+            templatesByMi.keys.toSeq.sortBy(id => (id.length, id)).flatMap { tag =>
+              templatesByMi(tag).sortBy(t => t.name).flatMap(t => t.primaryReads)
+            }
+          }
+          .seq
+          .iterator
```

Comment:
never sure if this is needed, but here we are
```suggestion
          .seq
          .iterator
```